app/api/csv/arrangeThesauri.ts

```diff
@@ -0,0 +1,180 @@
+  source: Error;
+  row: CSVRow;
+  index: number;
+
+  constructor(source: Error, row: CSVRow, index: number) {
+    super(source.message);
+    this.source = source;
+    this.row = row;
+    this.index = index;
+  }
+}
+
+const separateSelectAndMultiselectThesauri = (
+  thesauriRelatedProperties: PropertySchema[] | undefined,
+  languages?: string[]
+): [{ [k: string]: string }, { [k: string]: string }] => {
+  const nameToThesauriIdSelects: { [k: string]: string } = {};
+  const nameToThesauriIdMultiselects: { [k: string]: string } = {};
+
+  thesauriRelatedProperties?.forEach(p => {
+    if (p.content && p.type) {
+      const thesarusID = p.content.toString();
+      if (p.type === propertyTypes.select) {
+        nameToThesauriIdSelects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdSelects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      } else if (p.type === propertyTypes.multiselect) {
+        nameToThesauriIdMultiselects[p.name] = thesarusID;
+        languages?.forEach(suffix => {
+          nameToThesauriIdMultiselects[`${p.name}__${suffix}`] = thesarusID;
+        });
+      }
+    }
+  });
+
+  return [nameToThesauriIdSelects, nameToThesauriIdMultiselects];
+};
+
+type ThesauriValueData = {
+  thesauriIdToExistingValues: Map<string, Set<string>>;
+  thesauriIdToNewValues: Map<string, Set<string>>;
+  thesauriIdToNormalizedNewValues: Map<string, Set<string>>;
+};
+
+const setupIdValueMaps = (allRelatedThesauri: ThesaurusSchema[]): ThesauriValueData => {
+  const thesauriIdToExistingValues = new Map();
+  const thesauriIdToNewValues = new Map();
+  const thesauriIdToNormalizedNewValues = new Map();
+
+  allRelatedThesauri.forEach(t => {
+    if (t._id) {
+      const id = t._id.toString();
+      thesauriIdToExistingValues.set(
+        id,
+        new Set(t.values?.map(v => normalizeThesaurusLabel(v.label)))
+      );
+      thesauriIdToNewValues.set(id, new Set());
+      thesauriIdToNormalizedNewValues.set(id, new Set());
+    }
+  });
+
+  return { thesauriIdToExistingValues, thesauriIdToNewValues, thesauriIdToNormalizedNewValues };
+};
+
+const handleLabels = (
+  id: string,
+  original: string,
+  normalized: string | null,
+  thesauriValueData: ThesauriValueData
+) => {
+  if (
+    normalized &&
+    !thesauriValueData.thesauriIdToExistingValues.get(id)?.has(normalized) &&
+    !thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.has(normalized)
+  ) {
+    thesauriValueData.thesauriIdToNewValues.get(id)?.add(original);
+    thesauriValueData.thesauriIdToNormalizedNewValues.get(id)?.add(normalized);
+  }
+};
+
+const syncSaveThesauri = async (
+  allRelatedThesauri: ThesaurusSchema[],
+  thesauriIdToNewValues: Map<string, Set<string>>
+) => {
+  for (let i = 0; i < allRelatedThesauri.length; i += 1) {
+    const thesaurus = allRelatedThesauri[i];
+    if (thesaurus?._id) {
+      const newValues: { label: string }[] = Array.from(
```

Comment:
@LaszloKecskes allRelatedThesauri has a ThesaurusSchema[] type, if you check what thesauri.get returns its actually a diferent type, which includes the _id, making the condition on line 110 redundant and ensuring you will always have an _id, there should be a better way to achieve this, but it has not been worked out yet.

Reply:
`newValues` type should be inferred by the compiler for you, no need to explicitly type.

Reply:
Done.